lib/ansible/modules/system/lvol_thin.py

```diff
@@ -0,0 +1,755 @@
+            'size': float(decimal_point.match(parts[1]).group(1)),
+            'free': float(decimal_point.match(parts[2]).group(1)),
+            'vg_extent_size': float(decimal_point.match(parts[3]).group(1))
+        }
+
+
+class Lvol(object):
+
+    name = ''
+    size = None
+    unit = 'm'
+    pool_lv = ''
+    vg = None
+    yes_opt = ""
+    module = None
+
+    def __init__(self, module, vg_name, lv_name, unit, opts, pvs):
+        self.vg = Vg(module, vg_name, unit)
+        self.module = module
+
+        lv_info = self.get_lv_info(vg_name, lv_name, unit)
+
+        self.name = lv_info['name']
+        self.pool_lv = lv_info['pool_lv']
+        self.size = lv_info['size']
+        self.unit = unit
+        self.opts = opts
+        self.pvs = pvs
+
+        self.yes_opt = self.get_yes_opt(module)
+
+    @classmethod
+    def exists(cls, module, vg_name, lv_name):
+        exists = False
+        lvs_cmd = module.get_bin_path("lvs", required=True)
+
+        rc, current_lv, err = module.run_command(
+            "%s -a %s/%s" % (lvs_cmd, vg_name, lv_name)
+        )
+
+        if rc == 0:
+            exists = True
+        elif rc == 5 and "Failed to find logical volume" in err:
+            exists = False
+        else:
+            module.fail_json(msg="Unexpected error", rc=rc, err=err)
+
+        return exists
+
+    @classmethod
+    def is_thin_pool(cls, module, vg_name, lv_name):
+        is_thin_pool = False
+        lvs_cmd = module.get_bin_path("lvs", required=True)
+
+        rc, current_lv, err = module.run_command(
+            "%s -a --noheadings --nosuffix -o lv_layout --separator ';' %s/%s" % (lvs_cmd, vg_name, lv_name)
+        )
+
+        if rc != 0:
+            module.fail_json(msg="Error", rc=rc, err=err)
+
+        parts = current_lv.strip().split(';')
+
+        if 'thin,pool' in parts[0]:
+            is_thin_pool = True
+
+        return is_thin_pool
+
+    @classmethod
+    def is_thin_lv(cls, module, vg_name, lv_name):
+        is_thin_lv = False
+        lvs_cmd = module.get_bin_path("lvs", required=True)
+
+        rc, current_lv, err = module.run_command(
+            "%s -a --noheadings --nosuffix -o lv_layout --separator ';' %s/%s" % (lvs_cmd, vg_name, lv_name)
+        )
+
+        if rc != 0:
+            module.fail_json(msg="Error", rc=rc, err=err)
+
+        parts = current_lv.strip().split(';')
+
+        if 'thin' in parts[0] and 'pool' not in parts[0]:
+            is_thin_lv = True
+
+        return is_thin_lv
+
+    @classmethod
+    def mkversion(cls, major, minor, patch):
+        return (1000 * 1000 * int(major)) + (1000 * int(minor)) + int(patch)
+
+    @classmethod
+    def get_lvm_version(cls, module):
+        ver_cmd = module.get_bin_path("lvm", required=True)
+        rc, out, err = module.run_command("%s version" % (ver_cmd))
+
+        if rc != 0:
+            return None
+
+        m = re.search("LVM version:\s+(\d+)\.(\d+)\.(\d+).*(\d{4}-\d{2}-\d{2})", out)
```

Comment:
it might not be relevant in this particular case but be aware to better handle LANG when searching in output. use `module.run_command_environ_update = dict(LANG='C', LC_ALL='C', LC_MESSAGES='C', LC_CTYPE='C')` early after `module` has been instantiated to get that covered.